Remove backward op for slow 3d transposed convolution

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69933

This PR drops the backward op for slow 3d transposed convolution. It replaces the op with a dispatch stub, and registers a single composite CPU kernel for all CPU arch types.

Differential Revision: [D33131343](https://our.internmc.facebook.com/intern/diff/D33131343)

jbschlosser committed:
Remove backward op for slow 3d transposed convolution

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33131343).

jbschlosser committed:
Update on "Remove backward op for slow 3d transposed convolution"


Differential Revision: [D33131343](https://our.internmc.facebook.com/intern/diff/D33131343)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33131343).

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10042,18 +10042,6 @@
-  python_module: nn
-  dispatch:
-    CPU: slow_conv_transpose3d_backward_out_cpu
-    CUDA: slow_conv_transpose3d_backward_out_cuda
```
we can assume that backends aren't overriding this op, right? Since other backends (e.g. XLA) override `convolution_overrideable`, and not any of the individual ops.

jbschlosser replied:
That's correct AFAIK. `convolution_overrideable` was created to be overridden for XLA. I'm sure it's still possible that someone somewhere is overriding the op, but I haven't seen anything internal or external.

bdhirsh commented on aten/src/ATen/native/NaiveConvolutionTranspose3d.h:
```diff
@@ -0,0 +1,163 @@
+namespace native {
+namespace {
+
+static inline void slow_conv_transpose3d_shape_check(
```
should this go in a cpp file instead of a header?

bdhirsh commented on aten/src/ATen/native/cpu/NaiveConvolutionTranspose3d.cpp:
```diff
@@ -0,0 +1,563 @@
+
+  // number of input & output planes and kernel size is indirectly defined by
+  // the grad_weight tensor
+  slow_conv_transpose3d_shape_check(
```
It looks like you factored out the shape check into that other header file, but I also see a bunch of other shape checking here. What's the split for (or is some of it duplicated)?

jbschlosser replied:
Hey good question! Sorry I updated the descriptions of some PRs to explain the split, but missed this one. Essentially, I want to register a CPU dispatch, but the `REGISTER_DISPATCH` macro for CPU kernels requires that the code be place into the `native/cpu` dir where is it recompiled it once per arch type. So I moved all the backward logic underneath `native/cpu` so I can call `REGISTER_DISPATCH`.

I've been talking with Richard about this, and he is rightfully concerned that new multiple compilation unnecessarily regresses build time and expands the binary size. I've been throwing around an idea of defining a new macro `REGISTER_ALL_CPU_DISPATCH` that registers the same kernel across all arch types to avoid both the recompilation and the need to split the logic as done here. Do you hav any thoughts on this idea?

jbschlosser committed:
Update on "Remove backward op for slow 3d transposed convolution"


**Note:** `REGISTER_DISPATCH` for the CPU kernel is only accessible from the `native/cpu` directory. So this PR splits `aten/src/ATen/native/NaiveConvolutionTranspose3d.cpp` into:
* (new file) `aten/src/ATen/native/NaiveConvolutionTranspose3d.h` (contains functions shared between forward and backward)
* `aten/src/ATen/native/NaiveConvolutionTranspose3d.cpp` (contains forward logic)
* (new file) `aten/src/ATen/native/cpu/NaiveConvolutionTranspose3d.cpp` (contains backward functions + `REGISTER_DISPATCH` call)

Once the forward op is removed as well, the first two can go away.

Differential Revision: [D33131343](https://our.internmc.facebook.com/intern/diff/D33131343)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33131343).

bdhirsh reviewed (approved):
LGTM!
